File: data_structures/trees/dart/binary_search_trees.dart
import 'traversal.dart';

class BinarySearchTreeNode {
  final int value;
  BinarySearchTreeNode left;
  BinarySearchTreeNode right;

  BinarySearchTreeNode(this.value, this.left, this.right);
}

void main() {
  BinarySearchTreeNode root = BinarySearchTreeNode(1, null, null);
  root.left = BinarySearchTreeNode(2, null, null);
  root.right = BinarySearchTreeNode(3, null, null);
  root.left.left = BinarySearchTreeNode(4, null, null);
  root.left.right = BinarySearchTreeNode(5, null, null);
  print("Created BST!");
  traverseInOrder(root);
  print("");
}

File: data_structures/trees/cpp/binary_search_trees.cpp
#include <iostream>
using namespace std;

struct BinarySearchTreeNode
{
    int value;
    struct BinarySearchTreeNode *left, *right;
    BinarySearchTreeNode(int value)
    {
        this->value = value;
        left = right = NULL;
    }
};

void traverse_in_order(struct BinarySearchTreeNode *root);

int main()
{
    /*
	       1
	      / \
	     2   3
	    / \
	   4   5
	*/

    struct BinarySearchTreeNode *root = new BinarySearchTreeNode(1);
    root->left = new BinarySearchTreeNode(2);
    root->right = new BinarySearchTreeNode(3);
    root->left->left = new BinarySearchTreeNode(4);
    root->left->right = new BinarySearchTreeNode(5);

    cout << "Created BST! " << endl;

    traverse_in_order(root);    // This will print 4 2 5 1 3
    cout << endl;

    return 0;
}

void traverse_in_order(struct BinarySearchTreeNode *root)
{
    if (root == NULL)
        return;

    traverse_in_order(root->left);
    cout << root->value << " ";
    traverse_in_order(root->right);
}